node/src/components/consensus/protocols/highway.rs

```diff
@@ -283,6 +283,20 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
                         sender,
                     ));
                 } else {
+                    // If unit is sent by a doppelganger, deactivate this instance of an active
+                    // validator. Continue processing the unit so that it can be added to the state.
+                    if let Some(swunit) = pvv.inner().signed_wire_unit() {
```

Comment:
I'd do this match inside of `highway_core`: `HighwayProtocol` shouldn't have to care about whether it's a unit or an endorsement. And I think now that we really should also be doing the same check for endorsements anyway, which could be hidden away from `HighwayProtocol`.

Reply:
With endorsements, it's even more probable that someone will send our own back at us. I will add those.

Reply:
Done in [`dfce379` (#636)](https://github.com/CasperLabs/casper-node/pull/636/commits/dfce379e39018a399abcaaa44ad462853f8b3c83)